### PR TITLE
#167920214 ES6 Babel Configuration 

### PR DESCRIPTION
**Pull request by wahabtobibello**
#### What does this PR do?
Set up es6 babel configuration

#### Description of Task to be completed?
- Activate es6 syntax in project using babel
- Add useful plugins for es-next features

#### How should this be manually tested?
- clone the repo with `git clone https://github.com/andela/firestar-backend`
- cd into code base with `cd firestar-backend`
- run `npm install` to install packages
- use `npm run build` to run babel transpile
- use `npm start` to run the production server using the result of the previous step
- use `npm run start:dev` to run the development server

#### Any background context you want to provide?
- The `@babel/polyfill` package has been deprecated and has been replaced with the core-js and `regenrator-runtime` packages. [Learn more](https://babeljs.io/docs/en/babel-polyfill)

#### What are the relevant pivotal tracker stories?
[#167920214](https://www.pivotaltracker.com/n/projects/2382004/stories/167920214)

#### Screenshots (if appropriate)
N/A

#### Questions:
N/A

### CODE REVIEW

**Review by conquext — CHANGES_REQUESTED** [review]
ES6: Prefer arrow syntax in function definitions in the following files (not necessarily exhaustive):
* /config/passport.js
* /routes/api/index.js e.g. router.use((err, req, res, next) => {})
* /routes/api/users.js

No (default) export in models, index.js, routes/api

**Comment by wahabtobibello** [comment]
> ES6: Prefer arrow syntax in function definitions in the following files (not necessarily exhaustive):
> 
> * /config/passport.js
> * /routes/api/index.js e.g. router.use((err, req, res, next) => {})
> * /routes/api/users.js
> 
> No (default) export in models, index.js, routes/api



@conquext The changes you requested are not part of the scope of my PR. They are to be implemented using eslint.

**Review by marusoft — APPROVED** [review]
Great job @wahaaabello

**Comment by conquext** [comment]
> > ES6: Prefer arrow syntax in function definitions in the following files (not necessarily exhaustive):
> > 
> > * /config/passport.js
> > * /routes/api/index.js e.g. router.use((err, req, res, next) => {})
> > * /routes/api/users.js
> > 
> > No (default) export in models, index.js, routes/api
> 
> @conquext The changes you requested are not part of the scope of my PR. They are to be implemented using eslint.

Okay. What about the module exports?

**Comment by wahabtobibello** [comment]
@conquext It is also not part of the scope of my PR.

I am to change from commonjs export to es6 export (which I have done) but I am not to add or remove exports.

export style in the `babel.config.js` has to be in commonjs

**Comment by conquext** [comment]
> @conquext It is also not part of the scope of my PR.
> 
> I am to change from commonjs export to es6 export (which I have done) but I am not to add or remove exports.
> 
> export style in the `babel.config.js` has to be in commonjs

> *PT Description*: Configure application to use Es6.
@rovilay can you please comment on this?

**Comment by rovilay** [comment]
> > @conquext It is also not part of the scope of my PR.
> > I am to change from commonjs export to es6 export (which I have done) but I am not to add or remove exports.
> > export style in the `babel.config.js` has to be in commonjs
> 
> > _PT Description_: Configure application to use Es6.
> > @rovilay can you please comment on this?

@wahaaabello the scope of this work includes configuring the app to use es6 and also update all files to es6 syntax.

**Comment by wahabtobibello** [comment]
@rovilay Are you referring to the es5 syntax in the `babel.config.js` file?

**Comment by wahabtobibello** [comment]
@rovilay Are we totally outlawing all use of the `function () {}` function declaration?

**Comment by wahabtobibello** [comment]
I would like to refer all stakeholders to this [article](https://medium.com/@lucasdavidferrero/dont-use-arrow-functions-when-you-use-mongoose-schema-method-190b79f1640c) as to why this may lead to a bug. Basically, arrow functions don’t bind the `this` keyword as the function expression does.

Which is why I decided not to make the switch to the es6 arrow syntax. I recommend that I should only be use in the right context.

The `babel.config.js` file on the other hand, is not transpile by babel itself thus it runs in the node context.

**Comment by rovilay** [comment]
> I would like to refer all stakeholders to this [article](https://medium.com/@lucasdavidferrero/dont-use-arrow-functions-when-you-use-mongoose-schema-method-190b79f1640c) as to why this may lead to a bug. Basically, arrow functions don’t bind the `this` keyword as the function expression does.
> 
> Which is why I decided not to make the switch to the es6 arrow syntax. I recommend that I should only be use in the right context.
> 
> The `babel.config.js` file on the other hand, is not transpile by babel itself thus it runs in the node context.

@wahaaabello yes understood for the `babel.config.js`, but other files??

**Comment by wahabtobibello** [comment]
@conquext I believe all other files have been accounted for.

**Comment by conquext** [comment]
> > > ES6: Prefer arrow syntax in function definitions in the following files (not necessarily exhaustive):
> > > 
> > > * /config/passport.js
> > > * /routes/api/index.js e.g. router.use((err, req, res, next) => {})
> > > * /routes/api/users.js
> > > 
> > > No (default) export in models, index.js, routes/api
> > 
> > 
> > @conquext The changes you requested are not part of the scope of my PR. They are to be implemented using eslint.
> 
> Okay. What about the module exports?

The exports of index.js, models/ and passport.js are not effected.

**Comment by wahabtobibello** [comment]
@conquext 
The files you mentioned don't have exports, do they? Also `models/index.js` doesn't exist or maybe we are not referencing the same code base?

- src/config/passport.js

```
import passport from 'passport';
import { Strategy as LocalStrategy } from 'passport-local';
import mongoose from 'mongoose';

const User = mongoose.model("User");

passport.use(
    new LocalStrategy(
        {
            usernameField: "user[email]",
            passwordField: "user[password]"
        },
        function(email, password, done) {
            User.findOne({ email: email })
                .then(function(user) {
                    if (!user || !user.validPassword(password)) {
                        return done(null, false, {
                            errors: { "email or password": "is invalid" }
                        });
                    }

                    return done(null, user);
                })
                .catch(done);
        }
    )
);
```

- src/index.js
```
import fs from 'fs';
import http from 'http';
import path from 'path';
import methods from 'methods';
import express from 'express';
import bodyParser from 'body-parser;'
import session from 'express-session';
import cors from 'cors';
import passport from 'passport';
import errorhandler from 'errorhandler';
import mongoose from 'mongoose';
import models from './models/User';
import routes from './routes';
import method_override from 'method-override';
import morgan from 'morgan';

// if (isProduction) {
//     mongoose.connect(process.env.MONGODB_URI);
// } else {
//     mongoose.connect("mongodb://localhost/conduit");
//     mongoose.set("debug", true);
// }

const isProduction = process.env.NODE_ENV === "production";

// Create global app object
const app = express();

app.use(cors());

// swagger config middlewares
app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDocument));

// Normal express config defaults
app.use(morgan("dev"));
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json());

app.use(method_override());
app.use(express.static(__dirname + "/public"));

app.use(
  session({
    secret: 'authorshaven',
    cookie: {maxAge: 60000},
    resave: false,
    saveUninitialized: false,
  })
);

if (!isProduction) {
    app.use(errorhandler());
}

app.use(routes);

/// catch 404 and forward to error handler
app.use(function(req, res, next) {
  const err = new Error('Not Found');
  err.status = 404;
  next(err);
});

/// error handlers

// development error handler
// will print stacktrace
if (!isProduction) {
  app.use(function(err, req, res, next) {
    console.log(err.stack);

    res.status(err.status || 500);

    res.json({
      errors: {
        message: err.message,
        error: err,
      },
    });
  });
}

// production error handler
// no stacktraces leaked to user
app.use(function(err, req, res, next) {
  res.status(err.status || 500);
  res.json({
    errors: {
      message: err.message,
      error: {},
    },
  });
});

// finally, let's start our server...
const server = app.listen(process.env.PORT || 3000, function() {
  console.log('Listening on port ' + server.address().port);
});
```

**Comment by conquext** [comment]
You have to export passport in passport.js and the mongoose.model("users", UserSchema) in model/User. In a node project, the only way to share javascript files is to export them, with the exception of the index file (which is not required by any other file) and environment variables that can be accessed via dotenv. So basically, there is no point of any file that is not serving as a module to be imported by other part(s) of the project.

**Comment by wahabtobibello** [comment]
@conquext 
I get the need to export the User model but there is no need to export `app` from `src/index.js` unless maybe we wanted to integrate testing to the project and I don't believe there is any need to export the passport package because I think it is a global field.

But even then, this is not in the scope of my PR. Please understand that I'm trying to change only what is required of me and keep to the scope of my task so I don't break anything.

**Comment by conquext** [comment]
> @conquext
> I get the need to export the User model but there is no need to export `app` from `src/index.js` unless maybe we wanted to integrate testing to the project and I don't believe there is any need to export the passport package because I think it is a global field.
> 
> But even then, this is not in the scope of my PR. Please understand that I'm trying to change only what is required of me and keep to the scope of my task so I don't break anything.

You are right, the app export is unnecessary, I have withdrawn it. 
Doing the right thing does not break the code, it rather improves it.

**Comment by wahabtobibello** [comment]
@conquext 
I appreciate the feedback but then 
> If It ain't broke, don't fix it.

I assure you the app runs on `npm run start:dev`
Kindly review.

**Comment by conquext** [comment]
> @conquext
> I appreciate the feedback but then
> 
> > If It ain't broke, don't fix it.
> 
> I assure you the app runs on `npm run start:dev`
> Kindly review.

I believe the code works fine in ES5 as well, the objective of this task is not merely for it to work but for the task to be carried out according to the specifications.

**Review by conquext — APPROVED** [review]

File: src/index.js [hunk 1]
```diff
@@ -1,18 +1,29 @@
...
+import models from './models/User';
+import routes from './routes';
+import method_override from 'method-override';
+import morgan from 'morgan';
+import swaggerUi from 'swagger-ui-express';
+import swaggerDocument from '../swagger.json';
+
+// if (isProduction) {
+//     mongoose.connect(process.env.MONGODB_URI);
+// } else {
```
Comment (rovilay) [inline src/index.js]: Good work. Please remove commented code.

Comment (wahabtobibello) [reply src/index.js]: @rovilay Done!

**Pull request merged**